TACC/Core-CMS
# Quick: Simplify CSS Build Config & Fix Warnings

# Overview

1. Convert _complex_ `postcss.config.js` config files to _simple_ `.postcssrc.yml` files.¹
2. Fix unsightly warnings during CSS build process.²

> ¹ Some bits of logic could be removed: one was ineffectual before this change, one was ineffectual after this change.
> ² The warnings were two instances of `[…] --global-color-warning--[…] is undefined and used without a fallback`.

# Related

- https://github.com/TACC/Core-CMS-Resources/pull/43

# Changes

- __New__: Refactor `postcss.config.js` to `.postcssrc.yml`
- _Minor_: Prevent warning by importing a tool in each place that needs it.
- _Minor_: Remove comment that is outdated by a fix.
- __Fix__: Move CSS variable usage to after their definition.
    - This is relevant only because of "freezing" the variables. If this native CSS is left as-is, it works fine and as intended.

# Testing

1. Build Core and Frontera CSS on `main` (using its submodule commit).
1. Build Core and Frontera CSS on `quick/simplify-css-processing` (using its submodule commit). 
1. Confirm that built files have no unexpected difference.

<details><summary>Detailed Steps for Testing Build Output</summary>

1. Checkout `main`:
    - `git checkout main`
1. Update `tacciste_custom`:
    - `git submodule update`
        _If nothing happens, that is okay—it just means it is already up to date._
1. Set `CUSTOM_ASSET_DIR=frontera-cms` in `.env`.
1. Build CSS: `npm run build`.
1. __Confirm there is *no* error output.__
1. Copy built CSS to temporary directory for comparison:*
    - `cp -r taccsite_cms/static/site_cms/css/build taccsite_cms/static/site_cms/css/build.test`
    - `cp -r taccsite_custom/frontera-cms/static/frontera-cms/css/build taccsite_custom/frontera-cms/static/frontera-cms/css/build.test`
1. Add/stage the new `build.test` directories (so you can compare them when they change):
    1. `git add taccsite_cms/static/site_cms/css/build.test`
    2. `cd taccsite_custom`
    3. `git add frontera-cms/static/frontera-cms/css/build.test`
    4. `cd ../`
1. Checkout `quick/simplify-css-processing`.
    - `git checkout quick/simplify-css-processing`
1. Update `tacciste_custom`:
    - `git submodule update`
        _If the diff of `taccsite_custom` shows `-dirty`, that is expected—you made changes inside intentionally by creating the `build.test` directory._
1. Build CSS: `npm run build`.
1. __Confirm there is *no* error output.__
1. Overwrite previously copied built CSS with copy of newly built CSS:\*†
    - `rsync -a --delete taccsite_cms/static/site_cms/css/build/ taccsite_cms/static/site_cms/css/build.test/`
    - `rsync -a --delete taccsite_custom/frontera-cms/static/frontera-cms/css/build/ taccsite_custom/frontera-cms/static/frontera-cms/css/build.test/`
1. Review diff of `build.test/`'s files in working directory to staging/index.
    - `git diff`
1. Ensure only "Expected Differences in Build Output" appear.

> \* This is done so you can later compare the changes between built files on two branches using `git diff` (n.b. you copy the `build` directory so it can be tracked with Git _because the `build` directory is ignored by Git_).
> † [Use `rsync` to avoid OSX `cp` issues.](https://stackoverflow.com/a/53349667/11817077)

</details>

<details><summary>Expected Differences in Build Output</summary>

- __`.../site.css`__ & __`site.tacc-search-bar.css`__:
    The `--global-color-success--[…]` vars are declared __after__ `--global-color-warning--[…]` vars.
- __`.../freeze_variables/site.tacc-search-bar.css`__:
    The contents are minified.
- __`frontera-cms`:`.../css/build.test/template.home.css`__
    The movement of `.s-article-preview […]x-untruncate--one-line{overflow:visible;white-space:normal}` from center of file to top of file. (This has been tested via https://github.com/TACC/Core-CMS-Resources/pull/43.)

> __Note__: The file change to `s-article-list.css` has _zero_ effect on the Core's build output, because—currently—only Frontera imports that code.

</details>

# Notes

This variable freezing may be used or improved for GH-191. This PR merely cleans it up and fixes warnings.
